hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java

```diff
@@ -174,11 +159,120 @@ private PartitionedMapTable createTable(
         }
     }
 
-    private static PartitionedMapTable emptyMap(String mapName) {
+    @SuppressWarnings("rawtypes")
+    private FieldsMetadata getHeapMapFields(MapServiceContext context, String name) {
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            // Resolve sample.
+            RecordStore<?> recordStore = partitionContainer.getExistingRecordStore(name);
+
+            if (recordStore == null) {
+                continue;
+            }
+
+            Iterator<Map.Entry<Data, Record>> recordStoreIterator = recordStore.iterator();
+
+            if (!recordStoreIterator.hasNext()) {
+                continue;
+            }
+
+            Map.Entry<Data, Record> entry = recordStoreIterator.next();
+
+            return getFieldMetadata(entry.getKey(), entry.getValue().getValue());
+        }
+
+        return FieldsMetadata.EMPTY_ERROR;
+    }
+
+    @SuppressWarnings("rawtypes")
+    private FieldsMetadata getHdMapFields(MapContainer mapContainer) {
+        if (!nodeEngine.getProperties().getBoolean(GLOBAL_HD_INDEX_ENABLED)) {
+            // Cannot resolve fields when concurrent indexes are disabled
+            return FieldsMetadata.HD_ERROR;
+        }
+
+        InternalIndex[] indexes = mapContainer.getIndexes().getIndexes();
+
+        if (indexes == null || indexes.length == 0) {
+            // Cannot resolve fields when the map doesn't have concurrent indexes
+            return FieldsMetadata.HD_ERROR;
+        }
+
+        InternalIndex index = indexes[0];
+
+        Iterator<QueryableEntry> entryIterator = index.getSqlRecordIterator();
```

Comment:
If we have `NULL` values in the index, the iterator will return an entry corresponding to it. I guess this is OK for further resolution logic?

Reply:
We do not read index keys here, only actual entries. Therefore, the resolution logic is not affected.